### DISCUSSION TITLE
support default value using `getServerState`

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/zustand#1174] [Q&A]
I see there's a field called `getServerState`, but I don't see any usage of this API

https://github.com/pmndrs/zustand/blob/49dcec03ee9cd20a338b4e5a8219ef60ad944f7a/src/react.ts#L38

If we support like this, that will help SSR on `next.js`
```ts
import { ssr } 'zustand/middleware'
const defaultValue = { ... }
create(
  ssr(set => ({ ... }), () => {...defaultValue})
)
```

Upstream: #1145

**Comment by dai-shi** [comment]
Thanks for asking!
We have `getServerState`, but don't know how to use it yet.
So thanks for your help. Once we figure it out, let's write a doc.

`getServerState` has nothing to do with vanilla store, thus only defined in `react.ts`.

I'm hoping something like this would work:
```js
const defaultValue = { ... }
const useFooStore = create((set => ({ ... }))
useFooStore.getSererState = () => defaultValue
```

How's it?

**Reply by himself65** [reply]
I wrote a prototype for it: https://github.com/himself65/zustand/commit/da6b70ee3f1618c0354d4ba310d52c5c28807f88
What do you think abotu this?

**Reply by dai-shi** [reply]
I mean I'd want to avoid it as a middleware, because it's not for vanilla store.
But, that doesn't mean to stop you to create a third-party package.

**Reply by dai-shi** [reply]
But, but maybe it's not that bad idea.

**Reply by dai-shi** [reply]
tbh, I'm not so experienced with ssr, and not even sure if this is a good approach. there would be many issue reporting on it. so, it's more comfortable for me to start it as a 3rd-party library. and, we will see.

**Reply by himself65** [reply]
but in your example, you call `create` first, and it already passes `getServerSnapshot = api.getState` into `useSyncExternalStore`, I think that might won't work

**Reply by dai-shi** [reply]
`create` doesn't invoke `useSyncExternalStore`. Would you please give it a try? I need to re-consider if it doesn't work...

**Reply by himself65** [reply]
> `create` doesn't invoke `useSyncExternalStore`. Would you please give it a try? I need to re-consider if it doesn't work...

Oh, my bad. your idea works. Let me try a PR for it

**Comment by laclance** [comment]
any updates?

**Comment by arashi-dev** [comment]
no updates? :)

I saw this document: https://github.com/pmndrs/zustand/blob/bf9d0922fe9292c2399b9594eafe9304016b52d9/docs/guides/initialize-state-with-props.md

and thought maybe I can make it easier:
https://gist.github.com/arashi-dev/a8d01d206f4c315faef5c888d33f03df

I'm not sure if I did it right. it would be great if any of you can check it and give some idea :)

**Reply by dai-shi** [reply]
Using `createStore`, `useStore` and Context seems like a good approach. I'm not sure if the typing is ideal. Basically, we should infer store types as much as possible instead of you dealing with it.

**Reply by arashi-dev** [reply]
@dai-shi 
thanks for the comment. I will keep working on it.

any news about the `getServerState`? I think a solution by the maintainer (you) is far better than someone like me creating a third-party library

**Reply by arashi-dev** [reply]
also, at first, before reading the context approach, I had a simple project and this is how I solved the issue:

```ts
const makeStore = () => create({...})
```

##### _App.tsx
```ts
const isMountedRef = useRef(false)
const [myStore] = useState(makeStore)

if(!isMountedRef.current) {
    myStore.getState().hydrate.handler(props.zustand.myStore)  
    isMountedRef.current = true
}
```

however, I created some other functionality to make it work for multiple stores automatically (something like the gist I created).
and it was working well

**Reply by arashi-dev** [reply]
the idea of `hydrate.handler()` is from `next-redux-wrapper` which we had to add a `HYDRATE` action and handle merging server-side and client-side states

**Reply by arashi-dev** [reply]
happy new year btw :)

**Reply by dai-shi** [reply]
> any news about the `getServerState`? 

I think it's probably correct from the implementation perspective.
Unfortunately, I'm not super familiar with SSR best practice.
We need someone to help on this.

**Comment by tushar1998** [comment]
@dai-shi @himself65 
The feature can be something like
- Default value should be able to set on the server and then on client we wont need loading state
OR
- The create functio should have the ability to take default value
```javascript
const useCountStore = create(getDefaultValue, (set)=>{
   // ctx is placeholder for any data you want from or the first parameter can a object handled by user
   const defaultValue = getDefaultValue(ctx); // { count: 10 }
   
    return {
        ...defaultValue,
        resetMethod: () => set({count: 0})
    }
})
```

**Reply by dai-shi** [reply]
I think the current implementation exposes the capability as much as possible.
It's just how you would use it. If we find a good practice, we can consider providing something from the library if anything.

> `getDefaultValue(ctx);`

How do you implement ctx?

btw, you can do `useCountStore.setState(defaultValue)`. Please try that.

**Reply by tushar1998** [reply]
@dai-shi 
so you are saying that if I do `useCountStore.setState(defaultValue)` in server component can I get values on client side 

ctx: can be anything that the user wants, it can include set as well

Also we should add API reference docs

**Reply by dai-shi** [reply]
No, it's something that can only work in client components.

**Reply by dai-shi** [reply]
Is your question related to https://github.com/pmndrs/zustand/discussions/2023 ?

**Reply by tushar1998** [reply]
@dai-shi 
Yes, I got idea from react-query where `queryClient.prefetchQuery` does api call and state get hydrated on client side 
I wanted this kind of feature in zustand so that browser LCP (Longest contentful paint) remains minimum to zero

**Reply by dai-shi** [reply]
https://twitter.com/ephemjs/status/1706601277088473553 Check out this thread.
I hope someone can try creating a practice. We may or may not be able to create a library functionality. Will see.

I also wonder if anyone tries this: https://github.com/pmndrs/zustand/discussions/1174#discussioncomment-3345303

**Reply by tushar1998** [reply]
@dai-shi 
Currently I am using [zustand + context provider](https://github.com/pmndrs/zustand/blob/bf9d0922fe9292c2399b9594eafe9304016b52d9/docs/guides/initialize-state-with-props.md) but it will be amazing if its available directly in zustand

Does jotai solves my issue 👀?

**Reply by dai-shi** [reply]
If you need to run components on server (aka SSR), context would be required anyway to isolate state between sessions.

Jotai might work better with context, but it's still client states. So, essentially there is the same difficulty in either case.

**Reply by tushar1998** [reply]
They have useHydratedState hook 

So I made a component out of it like `Hydrate` in react-query

```
export function HydrateAtoms({
  children,
  atoms,
}: PropsWithChildren<HydrateAtomsProps>) {
  useHydrateAtoms([...atoms])
  return children
}
```

**Reply by dai-shi** [reply]
useHydrateAtoms is technically same as running `useCountStore.setState(defaultValue)` just once.
```js
const initialized = useRef(false);
if (!initialized.current) {
  initialize.current = true;
  useCountStore.setState(defaultValue);
}
```

**Reply by tushar1998** [reply]
Yeah, I realized that later after comparing both.
But on ground 0 if the states wont get updated on client side I wont either need atom or zustand, react context is enough for just sharing the server state onto its children client components

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/discussions/1537#discussioncomment-7355980

I think we should move the discussion here.

> https://twitter.com/dai_shi/status/1716304792786747646 We had a discussion in Twitter/X, which changed my mind.
> 
> I don't think we should add a new option `delayHydration` which is suboptimal, and it somewhat conflicts with `skipHydration` option.
> 
> Instead, we should explore the use of `.getServerState()` to avoid the **React** hydration error. My understanding is that we don't need to modify Zustand code. It's how we use it. So, it will be for docs, once we figure it out.
> 
> Would anyone be able to help? I think we need a minimal repro first. cc @TkDodo @pawelblaszczyk5 @dbritto-dev

If @knajjars can't provide a reproduction, can anyone follow? cc @charkour

**Reply by charkour** [reply]
Here is a simple [reproduction of the issue and steps on how to solve this](https://stackblitz.com/edit/stackblitz-starters-yxevax?file=app%2Fpage.tsx).

**Reply by charkour** [reply]
@dai-shi, here is my full write up of the issue which is also found on the StackBlitz reproduction link with examples of proper `persist` using with Next.js. Pleaes let me know what questions or concerns you may have. Thank you

----

```ts
/**
 * Zustand Next.js Hydration Issue
 *
 * Description: When using zustand with an isomorphic JavaScript framework that does it's first rendering pass on the server and then hydrates on the client, hydration issues will occur when using client-based storage (such as localStorage or sessionStorage) resulting in DOM mismatching. Next.js fails gracefully by reverting to full client-side rendering if this occurs, but it is slower than serving a fully static or server rendered HTML page.
 *
 * Why does this occur?
 * When using client-based storage (such a localStorage from the Browser API's), the server has no idea what the client's persisted state is, so the server will use the zustand store's default state which is `default` in this example. Then, when the client hydrates, zustand will hydrate the persisted value from the client, which may differ form the server's default state; if this occurs, it will cause a hydration issue: DOM mismatch.
 *
 * Is this a problem with zustand persist?
 * No, zustand persist naively assumes the user will provide the correct default state when initializing the store. If the server and client _both_ access to the persisted state (for example, the state could be hosted in an external database), then the store will hydrate with the same, correct state on both the client and the server solving the hydration issue.
 *
 * How to solve this issue?
 * There are multiple ways to solve hydration issues with Next.js (or other isomorephic JavaScript frameworks) by implementing persisted stores keeping the locaiton of the persisted state in mind. If using client-state, such as localStorage or sessionStorage, the user of zustand must know that the store's state may _only_ be rendered on the client. If using Next.js this can be achieved with `next/dynamic`. If not using Next.js this can be achieved by using `skipHydration` and `UseBoundStore.persist.rehyderate`.
 *
 * Summary:
 * This issue is a result of the persisted zustand store not having access to client-persisted state and the zustand developer must accomidate the server and client state's mismatching.
 *
 * How to use the reproduction:
 * 1) click "Open in New Tab"
 * 2) In the new tab, click the button to set-up the initial store
 * 3) Refresh the page to have Next.js re-render the page
 * 4) Next.js will throw an error for the DOM Mismatch from the first `RenderState` component: `hydration-default`
 * 5) Inspect the Stackblitz terminal to see the server-rendered state, and the browser console to see the client state
 *
 * Documentation:
 * - Two Next.js work arounds are provided below, and a third can be found here: https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#usage-in-nextjs
 * - zustand persist middleware documentation can be found here: https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#usage-in-nextjs
 * - next/dynamic documentation is here: https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading
 *
 * Related Zustand Issues:
 * - https://github.com/pmndrs/zustand/discussions/1174
 * - https://github.com/pmndrs/zustand/discussions/1537
 * - https://github.com/pmndrs/zustand/issues/324
 * - https://github.com/pmndrs/zustand/issues/1145
 * - https://github.com/pmndrs/zustand/issues/938
 */
```

**Reply by charkour** [reply]
Perhaps we should update these [docs](https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#usage-in-nextjs) with more Next.js best practices.

**Reply by dai-shi** [reply]
Thanks for the investigation.
What I would like to explore as a new solution is the use of `getServerState`, which I hope to avoid `skipHydration`. I'm not yet sure if it's possible, but what do you think if it would be possible?

**Reply by TkDodo** [reply]
Looked into this a bit today. I can see that zustand has `api.getServerState`, and `getServerSnapshot` of `useSyncExternalStore` also reads from it, but it's not actually set somewhere, so we just fall back to `getState`:

https://github.com/TkDodo/zustand/blob/9857a676a94c81ec561f99a13d46d6f6f3376f84/src/react.ts#L68

now `getState` will contain the state hydrated from localStorage via the `persist` middleware, so I think what we would need to do is capture the `initialState` and add `getServerState` to the `api` returned from `createStore`.

I have a PR up here:
- https://github.com/pmndrs/zustand/pull/2277

(just to get a codesandbox preview up and running so I can validate against the minimal example). If that works, I think we would still need a test that combines SSR with persist ?

**Reply by dai-shi** [reply]
> but it's not actually set somewhere,

Yes, you must define it on your end (at least in the current design).

> via the persist middleware

What I avoided is to add `getServerState` concept in `vanilla`, because it's React-only things. And `persist` is in `vanilla`. This complicates things though.

**Reply by TkDodo** [reply]
> This complicates things though.

yes it does because as far as I can see, we actually don't have access to the `initialState` inside `createStore` when combined with `persist`. One way to fix this would be to have `persist` alter the api to add `getServerState`, which wouldn't be unreasonable imo. It's a client side feature that alters what `getState` returns, so for it to work with server rendering, it must somehow expose the original state. Right now, it "swallows" it.

**Reply by dai-shi** [reply]
Just saw https://github.com/pmndrs/zustand/pull/2277/files#r1438180754. If I understand it correctly that that approach works, it's pretty simple.

**Comment by charkour** [comment]
@himself65, thanks for reporting this! I was also confused that `getServerState` is defined but not implemented and has no documentation. We'll see about make a change to avoid confusion. 

@dai-shi, I propose we deprecate `getServerState` and remove it in v5 because it does not seem to be necessary. What do you think?

**Reply by dbritto-dev** [reply]
@charkour I'm down with you we shall deprecate `getServerState`

**Reply by himself65** [reply]
> thanks for reporting this! 

I reported this 1 year ago🤯

**Comment by dai-shi** [comment]
> Perhaps we should update these [docs](https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#usage-in-nextjs) with more Next.js best practices.

https://github.com/pmndrs/zustand/discussions/2259#discussioncomment-7886053

**Comment by sovetski** [comment]
any updates please?

**Reply by dai-shi** [reply]
It's finished a long time ago. The original PR is #2277.
Please open a new discussion for further questions.
